http-client/src/main/java/io/micronaut/http/client/netty/Http1ResponseHandler.java

```diff
@@ -0,0 +1,355 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.client.netty;
+
+import io.micronaut.http.body.CloseableByteBody;
+import io.micronaut.http.client.exceptions.ResponseClosedException;
+import io.micronaut.http.netty.body.AvailableNettyByteBody;
+import io.micronaut.http.netty.body.BodySizeLimits;
+import io.micronaut.http.netty.body.BufferConsumer;
+import io.micronaut.http.netty.body.StreamingNettyByteBody;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.ReferenceCountUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Inbound message handler for the HTTP/1.1 client. Also used for HTTP/2 and /3 through message
+ * translators.
+ *
+ * @author Jonas Konrad
+ * @since 4.7.0
+ */
+final class Http1ResponseHandler extends SimpleChannelInboundHandlerInstrumented<HttpObject> implements BufferConsumer.Upstream {
+    private static final Logger LOG = LoggerFactory.getLogger(Http1ResponseHandler.class);
+
+    private final ResponseListener listener;
+    private State state = State.BEFORE_RESPONSE;
+    private HttpResponse response;
+    private List<ByteBuf> buffered;
+    private StreamingNettyByteBody.SharedBuffer streaming;
+    private long demand;
+    private ChannelHandlerContext streamingContext;
+
+    public Http1ResponseHandler(ResponseListener listener) {
+        super(false);
+        this.listener = listener;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.read();
+    }
+
+    @Override
+    protected void channelReadInstrumented(ChannelHandlerContext ctx, HttpObject msg) throws Exception {
+        if (msg.decoderResult().isFailure()) {
+            ReferenceCountUtil.release(msg);
+            exceptionCaught(ctx, msg.decoderResult().cause());
+            return;
+        }
+
+        HttpContent content;
+        switch (state) {
+            case BEFORE_RESPONSE:
```

Comment:
This switch-enum makes more sense with `state == null` as BEFORE_RESPONSE.
All other cases are content, which means you can already cast it to `HttpContent`